### PR TITLE
Update to_gbq and read_gbq to pandas-gbq 0.5.0

### LINKED ISSUE #177: specify geographical location of table in pandas read_gbq
I am having trouble setting the geographical location in `pandas.read_gbq`.

In `google.cloud.bigquery` I can set the location to `asia-northeast1` the following:

```
from google.cloud import bigquery
bigquery_client = bigquery.Client(project_name)    
query_job = bigquery_client.query(query, location='asia-northeast1')

```
However, I fail to do this in `read_gbq`. The following seems not to  work:

```
from pandas_gbq import gbq    
dat = gbq.read_gbq(query, project_id=project_name, jobReference={'location':'asia-northeast1'}, dialect='standard')
```

### LINKED ISSUE #21627: Update `to_gbq` and `read_gbq` signatures for Pandas-GBQ version 0.5.0
[Pandas-GBQ 0.5.0](https://github.com/pydata/pandas-gbq/releases/tag/0.5.0) adds several new features, including support for the BigQuery Tokyo region via a new `location` parameter (https://github.com/pydata/pandas-gbq/issues/177).

Note: I'm working on a PR for this right now.

### PR DESCRIPTION
**Pull request by tswast**
Closes https://github.com/pydata/pandas-gbq/issues/177

Closes #21627

- [x] closes #xxxx
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

I've also verified that the docs build and render well with
```
python doc/make.py --single read_gbq
python doc/make.py --single DataFrame.to_gbq
```

Output from `scripts/validate_docstrings.py pandas.read_gbq`:

<details>

```

################################################################################
######################### Docstring (pandas.read_gbq)  #########################
################################################################################

Load data from Google BigQuery.

This function requires the `pandas-gbq package
<https://pandas-gbq.readthedocs.io>`__.

See the `How to authenticate with Google BigQuery
<https://pandas-gbq.readthedocs.io/en/latest/howto/authentication.html>`__
guide for authentication instructions.

Parameters
----------
query : str
    SQL-Like Query to return data values.
project_id : str, optional
    Google BigQuery Account project ID. Optional when available from
    the environment.
index_col : str, optional
    Name of result column to use for index in results DataFrame.
col_order : list(str), optional
    List of BigQuery column names in the desired order for results
    DataFrame.
reauth : boolean, default False
    Force Google BigQuery to re-authenticate the user. This is useful
    if multiple accounts are used.
private_key : str, optional
    Service account private key in JSON format. Can be file path
    or string contents. This is useful for remote server
    authentication (eg. Jupyter/IPython notebook on remote host).
auth_local_webserver : boolean, default False
    Use the `local webserver flow`_ instead of the `console flow`_
    when getting user credentials.

    .. _local webserver flow:
        http://google-auth-oauthlib.readthedocs.io/en/latest/reference/google_auth_oauthlib.flow.html#google_auth_oauthlib.flow.InstalledAppFlow.run_local_server
    .. _console flow:
        http://google-auth-oauthlib.readthedocs.io/en/latest/reference/google_auth_oauthlib.flow.html#google_auth_oauthlib.flow.InstalledAppFlow.run_console

    *New in version 0.2.0 of pandas-gbq*.
dialect : str, default 'legacy'
    SQL syntax dialect to use. Value can be one of:

    ``'legacy'``
        Use BigQuery's legacy SQL dialect. For more information see
        `BigQuery Legacy SQL Reference
        <https://cloud.google.com/bigquery/docs/reference/legacy-sql>`__.
    ``'standard'``
        Use BigQuery's standard SQL, which is
        compliant with the SQL 2011 standard. For more information
        see `BigQuery Standard SQL Reference
        <https://cloud.google.com/bigquery/docs/reference/standard-sql/>`__.
location : str, optional
    Location where the query job should run. See the `BigQuery locations
    documentation
    <https://cloud.google.com/bigquery/docs/dataset-locations>`__ for a
    list of available locations. The location must match that of any
    datasets used in the query.

    *New in version 0.5.0 of pandas-gbq*.
configuration : dict, optional
    Query config parameters for job processing.
    For example:

        configuration = {'query': {'useQueryCache': False}}

    For more information see `BigQuery REST API Reference
    <https://cloud.google.com/bigquery/docs/reference/rest/v2/jobs#configuration.query>`__.
verbose : None, deprecated
    Deprecated in Pandas-GBQ 0.4.0. Use the `logging module
    to adjust verbosity instead
    <https://pandas-gbq.readthedocs.io/en/latest/intro.html#logging>`__.

Returns
-------
df: DataFrame
    DataFrame representing results of query.

See Also
--------
pandas_gbq.read_gbq : This function in the pandas-gbq library.
pandas.DataFrame.to_gbq : Write a DataFrame to Google BigQuery.

################################################################################
################################## Validation ##################################
################################################################################

Errors found:
	No examples section found
```

</details>

Output from `scripts/validate_docstrings.py pandas.DataFrame.to_gbq`:

<details>

```
################################################################################
##################### Docstring (pandas.DataFrame.to_gbq)  #####################
################################################################################

Write a DataFrame to a Google BigQuery table.

This function requires the `pandas-gbq package
<https://pandas-gbq.readthedocs.io>`__.

See the `How to authenticate with Google BigQuery
<https://pandas-gbq.readthedocs.io/en/latest/howto/authentication.html>`__
guide for authentication instructions.

Parameters
----------
destination_table : str
    Name of table to be written, in the form ``dataset.tablename``.
project_id : str, optional
    Google BigQuery Account project ID. Optional when available from
    the environment.
chunksize : int, optional
    Number of rows to be inserted in each chunk from the dataframe.
    Use ``None`` to load the dataframe in a single chunk.
reauth : bool, default False
    Force Google BigQuery to re-authenticate the user. This is useful
    if multiple accounts are used.
if_exists : str, default 'fail'
    Behavior when the destination table exists. Value can be one of:

    ``'fail'``
        If table exists, do nothing.
    ``'replace'``
        If table exists, drop it, recreate it, and insert data.
    ``'append'``
        If table exists, insert data. Create if does not exist.
private_key : str, optional
    Service account private key in JSON format. Can be file path
    or string contents. This is useful for remote server
    authentication (eg. Jupyter/IPython notebook on remote host).
auth_local_webserver : bool, default False
    Use the `local webserver flow`_ instead of the `console flow`_
    when getting user credentials.

    .. _local webserver flow:
        http://google-auth-oauthlib.readthedocs.io/en/latest/reference/google_auth_oauthlib.flow.html#google_auth_oauthlib.flow.InstalledAppFlow.run_local_server
    .. _console flow:
        http://google-auth-oauthlib.readthedocs.io/en/latest/reference/google_auth_oauthlib.flow.html#google_auth_oauthlib.flow.InstalledAppFlow.run_console

    *New in version 0.2.0 of pandas-gbq*.
table_schema : list of dicts, optional
    List of BigQuery table fields to which according DataFrame
    columns conform to, e.g. ``[{'name': 'col1', 'type':
    'STRING'},...]``. If schema is not provided, it will be
    generated according to dtypes of DataFrame columns. See
    BigQuery API documentation on available names of a field.

    *New in version 0.3.1 of pandas-gbq*.
location : str, optional
    Location where the load job should run. See the `BigQuery locations
    documentation
    <https://cloud.google.com/bigquery/docs/dataset-locations>`__ for a
    list of available locations. The location must match that of the
    target dataset.

    *New in version 0.5.0 of pandas-gbq*.
progress_bar : bool, default True
    Use the library `tqdm` to show the progress bar for the upload,
    chunk by chunk.

    *New in version 0.5.0 of pandas-gbq*.
verbose : bool, deprecated
    Deprecated in Pandas-GBQ 0.4.0. Use the `logging module
    to adjust verbosity instead
    <https://pandas-gbq.readthedocs.io/en/latest/intro.html#logging>`__.

See Also
--------
pandas_gbq.to_gbq : This function in the pandas-gbq library.
pandas.read_gbq : Read a DataFrame from Google BigQuery.

################################################################################
################################## Validation ##################################
################################################################################

Errors found:
	No returns section found
	No examples section found
```

</details>

### CODE REVIEW

**Comment by tswast** [comment]
Travis failure appears to be unrelated to this change.

File: pandas/io/gbq.py [hunk 1]
```diff
@@ -62,6 +54,16 @@ def read_gbq(query, project_id=None, index_col=None, col_order=None,
         Service account private key in JSON format. Can be file path
         or string contents. This is useful for remote server
         authentication (eg. Jupyter/IPython notebook on remote host).
+    auth_local_webserver : boolean, default False
+        Use the `local webserver flow`_ instead of the `console flow`_
+        when getting user credentials.
+
+        .. _local webserver flow:
```
Comment (jreback) [inline pandas/io/gbq.py]: i think these break linting, need a ``#noqa`` I think

File: pandas/core/frame.py [hunk 2]
```diff
@@ -1102,37 +1102,27 @@ def to_dict(self, orient='dict', into=dict):
...
+        <https://pandas-gbq.readthedocs.io/en/latest/howto/authentication.html>`__
+        guide for authentication instructions.
 
         Parameters
         ----------
         destination_table : str
-            Name of table to be written, in the form 'dataset.tablename'.
-        project_id : str
-            Google BigQuery Account project ID.
+            Name of table to be written, in the form ``dataset.tablename``.
```
Comment (jreback) [inline pandas/core/frame.py]: I think needs to be single backticks

File: doc/source/whatsnew/v0.24.0.txt [hunk 3]
```diff
@@ -18,6 +18,8 @@ Other Enhancements
 - :func:`Series.mode` and :func:`DataFrame.mode` now support the ``dropna`` parameter which can be used to specify whether NaN/NaT values should be considered (:issue:`17534`)
 - :func:`to_csv` now supports ``compression`` keyword when a file handle is passed. (:issue:`21227`)
 - :meth:`Index.droplevel` is now implemented also for flat indexes, for compatibility with MultiIndex (:issue:`21115`)
+- :func:`to_gbq` and :func:`read_gbq` signature and documentation updated to
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.txt]: can you add here a link to the pandas-gbq docs whatsnew for 0.5.0?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/frame.py (hunk 2 above)
Comment (tswast) [reply pandas/core/frame.py]: Two backticks is [code font in Sphinx RST](http://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#inline-markup), which is what I want.

File: doc/source/whatsnew/v0.24.0.txt (hunk 3 above)
Comment (tswast) [reply doc/source/whatsnew/v0.24.0.txt]: Added link to https://pandas-gbq.readthedocs.io/en/latest/changelog.html#changelog-0-5-0

File: pandas/io/gbq.py (hunk 1 above)
Comment (tswast) [reply pandas/io/gbq.py]: Adding `# noqa` breaks the link. Sphinx generates a link like `pandas/doc/build/html/generated_single/pandas.read_gbq.html#noqahttp://google-auth-oauthlib.readthedocs.io/en/latest/reference/google_auth_oauthlib.flow.html#google_auth_oauthlib.flow.InstalledAppFlow.run_local_server` when I add this.

I do not get any lint errors when I run `git diff upstream/master -u -- "*.py" | flake8 --diff` and `scripts/validate_docstrings.py pandas.read_gbq` also passes.

**Review by tswast — COMMENTED** [review]
Updated with link to Pandas-GBQ changelog and rebase on master.

**Comment by tswast** [comment]
Looks like Travis and Circle are both happy now that I rebased.

**Comment by jorisvandenbossche** [comment]
I am not really familiar with the gbq code, but, did we consider making this just a passthrough of `*args, **kwargs` ? Then it would automatically work with the installed pandas_gbq

**Comment by tswast** [comment]
@jorisvandenbossche I considered (and even coded) that in https://github.com/pandas-dev/pandas/pull/20564

The development experience is nicer (IDEs understand them better) without using `**kwargs`. It is somewhat rare to add new arguments to Pandas-GBQ, though I do have a couple changes in mind to streamline auth https://github.com/pydata/pandas-gbq/issues/161.

The `location` field is special because it's a new field on the BigQuery API as opposed to a new field in the BigQuery API resources.

**Comment by jreback** [comment]
i think writing out the kwargs is nice from a usability pov - we do this for other apis as well

**Comment by jorisvandenbossche** [comment]
Yep, no problem, was just wondering!

**Review by jreback — APPROVED** [review]

**Pull request merged**